### ISSUE TITLE
Produktleser verschwindet wenn eine Variante aktiviert wird - PHP 5.6.8

### ISSUE THREAD

**Issue by akroii** [isotope/core#1452]
Hallo,
wenn man unter PHP 5.6.8 den Isotope betreibt, verschwindet der Produktleser ohne eine Meldung im Frontend, sobald ich eine Variante aktiviere.
Was passiert genau?
Produktliste ->Produktleser (hier wird noch alles vollständig angezeigt) -> Variante aus SelectMenü auswählen -> "Lade Produktdaten..." -> Produkt verschwindet (Siehe Anhang)

Ich habe einmal im iso_list_default und iso_reader_default einmal var_dump($this) ausgeführt um zusehen was denn überhaupt mit den Daten passiert....dabei erhalte ich aber keine Ausgabe. Er springt also komplett raus aus dem Template. Error.log ist leer.

Ich benutze Contao 3.4.4 mit Isotope 2.2.4

**Comment by tsarma** [comment]
Ditto here: https://github.com/isotope/core/issues/1451

**Comment by akroii** [comment]
But in my case the problem appears after i clicked the variant in the selectmenu

**Comment by akroii** [comment]
after i tried some various things, i can circumscribe where the problem is

When i click in the productreader "show 404 page" i got this in frontend:
![bildschirmfoto 2015-06-11 um 17 06 55 kopie](https://cloud.githubusercontent.com/assets/8830861/8110737/3fcac0d0-105e-11e5-8b29-8f2b98ec994a.jpg)

**Comment by tsarma** [comment]
Now I realize the problem I have described and yours is little different. 

I am trying to list all the variants of a product along with the product in product-listing module. And the main product has to show in the primary list, not the one with lowest price from the variants.

After reading your above comments, I hope you could now solve yours problem.

**Comment by akroii** [comment]
I add an image.
![bug](https://cloud.githubusercontent.com/assets/8830861/8164459/72516ab2-1389-11e5-85ff-56488ef77d1a.gif)

**Comment by sheeep** [comment]
The problem is [this exact line](https://github.com/isotope/core/blob/master/system/modules/isotope/assets/js/isotope.js#L173):

```js
div.innerHTML = txt;
```

If the response you get (a `form` tag), is not trimmed and contains a leading break `↵` sign, the next line in the script will return a text node, instead of the `form` tag.

```js
var newForm = div.firstChild;
```

... which will then replace the old form tag.

```js
form.parentNode.replaceChild(newForm, form);
```

So you basically fill the node with an empty text node instead of the proper `form` tag.

For a quick fix remove every whitespace before the form tag in all the templates found [here](https://github.com/isotope/core/tree/master/system/modules/isotope/templates/isotope). I'm not exactly sure if all of them are affected in a way or another.

For isotope in general: A proper way would be to either
* Trim the response before sending it back to the browser (if you really want to rely on vanilla-js node generation)
* Trim the response after sending it back to the browser (`txt = txt.trim()`)
* Use a Mootools-y way of creating elements like described [here](http://mootools.net/more/docs/1.5.1/Element/Elements.From), not sure they contain a fix for this problem though.

I'm happy to provide a PR if needed, just let me know which way is preferred.

**Comment by aschempp** [comment]
Great find! The best solution would probably be to use `div.children[0` as that would not return text nodes ;-)

**Comment by sheeep** [comment]
Seems like Mootools is [pretty much doing the same](https://github.com/mootools/mootools-more/blob/master/Source/Element/Elements.From.js#L39):

```js
return /* [...] */ new Element('div')).set('html', text).getChildren();
```

Where `getChildren` seems to [omit text nodes](http://stackoverflow.com/questions/2173929/mootools-getchildren-including-text-nodes) as well.

[For reference](https://developer.mozilla.org/de/docs/Web/API/ParentNode/children): 
`
Internet Explorer 6, 7 and 8 supported it, but erroneously includes Comment nodes.
`

If you ask me, the correct way would probably be to just use `Element.children`. And/or [`trim`](https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/String/Trim)  on the browser site. IE<9 needs a pollyfill anyway though.

**Comment by akroii** [comment]
@aschempp @sheeep  Thanks a lot!
Can i use div.children[0]; as a working hotfix?

**Comment by bytehead** [comment]
@akroii it's enough to just remove any newlines before the `<form>`-tag in `iso_reader_default.html5` and `iso_list_default.html5`.

**Comment by akroii** [comment]
@bytehead it doesn't work to remove whitespace/newlines only

**Comment by akroii** [comment]
@bytehead did you try it?

When i add: ```div.children[0];``` i got this in console in ```isotope.min.js```: 
```
Uncaught TypeError: Failed to execute 'replaceChild' on 'Node': parameter 1 is not of type 'Node'.Request.HTML.onSuccess @ isotope.min.js:1(anonymous function) @ edd83b8736a3.js:4(anonymous function) @ edd83b8736a3.js:4Array.implement.each @ edd83b8736a3.js:4Events.Class.fireEvent @ edd83b8736a3.js:4r.extend.$owner @ edd83b8736a3.js:4n.Request.Class.onSuccess @ edd83b8736a3.js:6r.extend.$owner @ edd83b8736a3.js:4Request.HTML.Class.success @ edd83b8736a3.js:6r.extend.$owner @ edd83b8736a3.js:4n.Request.Class.onStateChange @ edd83b8736a3.js:6r.extend.$owner @ edd83b8736a3.js:4```

**Comment by bytehead** [comment]
It works on my system, if you remove the leading newlines in the templates.

**Comment by akroii** [comment]
ok, it works. It is important to remove not only the newlines, remove the spaces before each div is necessary too.

**Comment by seaneble** [comment]
This is a nasty bug. I just invested half an hour to debug the JS and then gladly found this issue.
Isotope itself does not really have a problem here, but as soon as the custom reader template includes a blank line (which can easily happen and there is no way to stop that), the shop is broken.
Could you please fix the JS so that another method to retrieve the HTML structure from the server response is used? Thanks :smile:

**Comment by aschempp** [comment]
Fixed in aa7e35401ceddc2b3de464c395992e720e1f5621
